src/test/java/games/strategy/engine/delegate/DelegateExecutionManagerTest.java

```diff
@@ -0,0 +1,19 @@
+package games.strategy.engine.delegate;
+
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import org.junit.jupiter.api.Test;
+
+public final class DelegateExecutionManagerTest {
+  private final DelegateExecutionManager delegateExecutionManager = new DelegateExecutionManager();
+
+  @Test
+  public void enterDelegateExecution_ShouldThrowExceptionWhenDelegateExecutingOnCurrentThread() {
```

Comment:
I'm a bit concerned it may become eventually become difficult over time to discern the reasoning behind this test. I think some natural questions would be raised:
-  why is it important that delegate should not execute on current thread?
- 'current' thread is relative to context, what is the 'current' thread? Are we talking swing thread or maybe game client thread?
- in which cases do we not get an exception? 
- As there are no other test cases, can we be sure this is not a tautology test? (Giving a positive test case that shows no exception would document this, alternatively a comment on why we know we would not get an exception otherwise, or why we test this specific case should help)

Reply:
> why is it important that delegate should not execute on current thread?

A delegate should not execute on the current thread _while another delegate is executing on the current thread._  Presumably bad things can happen if two delegates are actively executing on the same thread.

My guess is the original designer of the delegate framework added this check to help devs detect the case where they incorrectly nest execution of a delegate from within another delegate instead of queuing the new delegate's execution to begin after the current delegate's execution ends.

> 'current' thread is relative to context, what is the 'current' thread? Are we talking swing thread or maybe game client thread?

There doesn't appear to be any constraints upon which thread a delegate can execute.  So "current" simply means the thread where you're trying to execute a delegate. :smile:

> in which cases do we not get an exception?

When serializing their execution on the same thread.  I'll add a test for that.

Reply:
1e8acee75

I also added a test to demonstrate concurrent delegate execution is permitted on different threads in bd069b916.